[content-visibility] Improve skipped content naming in RenderStyle

rwlbuis opened the pull request:
#### b8e28937f4ed4c2f98458b185f7b462bc1c1a18e
<pre>
[content-visibility] Improve skipped content naming in RenderStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=267817">https://bugs.webkit.org/show_bug.cgi?id=267817</a>

Reviewed by Tim Nguyen.

In hindsight skippedContentReason returning a std::optional seems to make
things less readable, especially since we test for has_value() a lot instead
of relying on the implicit bool conversion, so remove it, and rename skippedContentReason
to effectiveContentVisibility.
Also add a hasSkippedContent method as a more readable replacement
for the has_value() checks.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasFocusableStyle const):
(WebCore::Element::checkVisibility):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::updateDescendantDependentFlags):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isSkippedContent const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresLayout):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::effectiveContentVisibility const):
(WebCore::RenderStyle::hasSkippedContent const):
(WebCore::RenderStyle::skippedContentReason const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setEffectiveContentVisibility):
(WebCore::RenderStyle::setSkippedContentReason): Deleted.
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Canonical link: <a href="https://commits.webkit.org/273276@main">https://commits.webkit.org/273276@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/655fc3141718e74d70c239e2606e1febd2f7b8b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34963 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13834 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37023 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37693 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31566 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36115 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16229 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10912 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30508 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35500 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11726 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31164 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10266 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10332 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31253 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38945 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31789 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31582 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36353 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10431 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8349 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34332 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12240 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8011 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10969 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11303 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5482,11 +5482,11 @@ bool Element::checkVisibility(const CheckVisibilityOptions& options)
+        if (style->effectiveContentVisibility() != reason)
             return false;
 
         // skippedContentReason() includes the skipped content root, so we query the parent to make sure roots are not considered as skipped.
```
Can you update this comment?

rwlbuis replied:
Well spotted, fixed.

nt1m commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -715,7 +715,8 @@ class RenderStyle {
 
-    inline std::optional<ContentVisibility> skippedContentReason() const;
+    inline ContentVisibility effectiveContentVisibility() const;
+    inline bool hasSkippedContent() const;
```
Maybe a comment explaining the return values would be useful. It's not super obvious what the first one returns in the `auto` case.

rwlbuis replied:
True, I added comments.

nt1m commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -715,7 +715,12 @@ class RenderStyle {
-    inline std::optional<ContentVisibility> skippedContentReason() const;
+    // effectiveContentVisibility will return ContentVisibility::Hidden if a content-visibility: hidden
+    // subtree (overriding content-visibility: auto at all times), ContentVisibility::Auto in a
+    // content-visibility: auto subtree, and ContentVisibility::Visible otherwise.
```
I think it's important to clarify that in the `auto` case, we only return `auto` when the content is not relevant (or skipped).

rwlbuis replied:
Good point, fixed.

rwlbuis committed:
[content-visibility] Improve skipped content naming in RenderStyle
https://bugs.webkit.org/show_bug.cgi?id=267817

Reviewed by Tim Nguyen.

In hindsight skippedContentReason returning a std::optional seems to make
things less readable, especially since we test for has_value() a lot instead
of relying on the implicit bool conversion, so remove it, and rename skippedContentReason
to effectiveContentVisibility.
Also add a hasSkippedContent method as a more readable replacement
for the has_value() checks.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasFocusableStyle const):
(WebCore::Element::checkVisibility):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::updateDescendantDependentFlags):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isSkippedContent const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresLayout):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::effectiveContentVisibility const):
(WebCore::RenderStyle::hasSkippedContent const):
(WebCore::RenderStyle::skippedContentReason const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setEffectiveContentVisibility):
(WebCore::RenderStyle::setSkippedContentReason): Deleted.
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Canonical link: https://commits.webkit.org/273276@main
